### DISCUSSION TITLE
Who-Is Service response : Device Object's Instance number and Address

### DISCUSSION THREAD

**Discussion by rprronad** [ChristianTremblay/BAC0#418] [General]
Who is allocating the **Device Object's Instant number** and **address** to the BACnet device by sending **Who-Is service** from the client device and also Who can able use(_Only client or Both client and server_) the Who-Is service in the BACnet network?

**Comment by bbartling** [comment]
can you give more context to your use case?

if you are setting up inside a building for the first time imitating this command I think its the devices inside the building that responding. The devices BACnet Instance ID and address is configured by the vendor that initially set it up and each device needs to be unique.

_**I think**_ BAC0 does a `who-is` or something like that to the nature on an interval so it understands the local environment.

**Reply by rprronad** [reply]
I'm using "Yabe" simulator for checking the Who - Is service , so can you confirm from the below attached file, where I'm getting response as a IP address and Object instance number in the console. 

Can i get a clarification on **Object instance number**, 
Who is assigning this number for BACnet device? 
Is it a unique number in whole BACnet network device? 

![Who-Is service](https://github.com/ChristianTremblay/BAC0/assets/124079051/e9990b0d-1038-4841-8540-d77c7894c03a)

**Reply by bbartling** [reply]
@ChristianTremblay correct me if i am wrong on these statements.

It looks like you have one device responding to your who-is which is BACnet instance ID 3057042

if I ran that in my building about 50 devices would populate.  a mix of Johnson Control devices and Trane and a rooftop unit by another vendor that runs a Tridium appliance inside it. All of the IP addresses (or MSTP devices) are all setup by the controls contractors that setup the equipment long before I arrived at the building...and typically the IP addresses are allocated by the IT department for the building.

Hopefully BACnet instance ID's are managed too if multiple vendors are at a building they all need to coordinate together on not duplicating an instance ID accidently. Those all need to be unique. For example if a Trane device and a JCI device had the same instance ID funny stuff would happen it could even make devices go offline or not respond to the who-is.

So the short story is the instance ID's are all setup by the controls contractors setting up the junk in the building...and if you need to change an instance ID you may have to hire them or get the software to do so.

Also typically only run the `who-is` not very often...like its a global request and if there is a lot of devices on a twisted pair MSTP networks I think it can create a lot of traffic like 0.01 sleep in between the who-is requests seems too fast, like it maybe creating ALOT networks traffic jams. Maybe adjust that to like 10 seconds or 5 seconds in IMHO

**Reply by rprronad** [reply]
Thanks @bbartling,

From the below attached file, I came to know that Instance number has been generated for the particular BACnet device, so can you confirm for particular case, if two vendor using this BAC0 library & developed some protocol converters means, at time it will give same Instance ID for vendor device and then it will go offline as you said earlier.

So How can it be identified and resolve the conflict of Instance number ? 

![instance number](https://github.com/ChristianTremblay/BAC0/assets/124079051/a66aa426-0efc-4a2a-a795-0d2bd5e267b7)

**Reply by bbartling** [reply]
when i make an app like:

`bacnet = BAC0.lite()`

if i need it to be custom IDs or ports, etc... I make `bacnet` var with passing in something like this below. I am experimenting with dual NIC card Nano Pi device where the internet connection is one NIC and the other NIC is for the OT LAN that requires a static IP for the buildings BACnet network:

```
# BACnet NIC setup:
BACNET_INST_ID = 3056672
IP_ADDRESS = "192.168.0.109"
SUBNET_MASK_CIDAR = 24
PORT = "47808"
BBMD = None

bacnet = BAC0.lite(ip=IP_ADDRESS, port=PORT , mask=SUBNET_MASK_CIDAR, deviceId=BACNET_INST_ID, bbmdAddress=BBMD)
```

you dont need to pass in all those params, if you just need a custom `deviceID` you can just pass that one in or set the others as `None`

I typically set my device ID's real high with the assumptions that the people setting up the stuff inside the buildings set devices IDs real low...and it generally works. At least when I was in the field setting up the stuff I would just start at device 1 and increment up by 1 for each embedded device I was setting up in the temperature control system where it was rare if it were the last device was over the value of 100. A 100+ BACnet devices inside a building is more than likely a large building...so the field guys usually make device ID's low and if you go high shooting at the hip your generally good to go

**Reply by rprronad** [reply]
@bbartling Thanks, Gave me good information.

and now Can you suggest me How to do `I/O data mapping table` for BACnet to Modbus Protocol converter ? Give some `examples `as well.

**Reply by bbartling** [reply]
are building something custom for a specific building or a very generic application that would work in any building?

is your app to be a "server" where people on some external platform write to it like via Modbus where then there would be a BACnet API on the other side to see translated values? OR would it be your just doing something custom on some random building... read Modbus values, then write to the BACnet system for some translated values of the Modbus sensor..

**Reply by rprronad** [reply]
I'm developing s/w for "very generic application that would work in any building" and that Protocol converter is to translate the Data from Modbus registers into BACnet things.

**Comment by bbartling** [comment]
So if you wanted to make a gateway for Modbus to BACnet this may get you started. I have only tinkered with Modbus and dont know a whole lot about.

This code below the app reads a modbus sensor value (its a power meter) every 60 seconds and then it also acts as a discoverable BACnet device (server) where then the modbus sensor can be visible to the BACnet network if other devices or control systems request that value from the BACnet server. 

I dont honestly know a ton about Modbus and when I was making my app I had to get some help on stackoverflow to making the Modbus client read request work...but that code below did work on a power meter inside a building at one point in time the Modbus registry stuff could be very unique to the device

Good luck : )

```
import asyncio
from datetime import datetime, timezone, timedelta
import threading
import logging
from BAC0.core.devices.local.models import analog_value
import BAC0

from pymodbus.client import ModbusTcpClient
from pymodbus.constants import Endian
from pymodbus.payload import BinaryPayloadDecoder

# Constants for BACnet NIC setup
IP_ADDRESS = "192.168.0.109"
SUBNET_MASK_CIDAR = 24
PORT = "47820"
BBMD = None
BACNET_INST_ID = 3056672

# Constants for the Modbus meter sensor
MODBUS_METER_ADDRESS = '192.168.0.111'
MODBUS_METER_PORT = 502
MODBUS_INPUT_REG = 17

# BACnetApp class for BACnet setup
class BACnetApp:
    @classmethod
    async def create(cls):
        self = BACnetApp()
        self.bacnet = await asyncio.to_thread(BAC0.lite, ip=IP_ADDRESS, port=PORT, mask=SUBNET_MASK_CIDAR, deviceId=BACNET_INST_ID, bbmdAddress=BBMD)
        self.building_meter = -1.0  # Default power value
        _new_objects = self.create_objects()
        _new_objects.add_objects_to_application(self.bacnet)
        logging.info("BACnet APP Created Success!")
        return self

    # create one discoverable BACnet point on the your 
    # BACnet server app
    def create_objects(self):
        _new_objects = analog_value(
            name="power-level",
            description="Writeable point for electric meter reading",
            presentValue=-1.0, # default data
            is_commandable=False,
        )
        return _new_objects

    async def update_bacnet_api(self, value):
        electric_meter_obj = self.bacnet.this_application.get_object_name("power-level")
        electric_meter_obj.presentValue = value
        print("BACnet API updated to:", electric_meter_obj.presentValue)

    async def keep_baco_alive(self):
        counter = 0
        while True:
            counter += 1
            await asyncio.sleep(0.01)
            if counter == 100:
                counter = 0
                async with self.server_check_lock:
                    await self.update_bacnet_api(self.building_meter)

    async def modbus_meter_reader(self):

        try:
            client = ModbusTcpClient(MODBUS_METER_ADDRESS,
                                     port=MODBUS_METER_PORT)
            
            result = client.read_input_registers(500,2,units=1)
            print(result.registers)
            decoder = BinaryPayloadDecoder.fromRegisters(result.registers, 
                                                         byteorder=Endian.Big)
            data = decoder.decode_32bit_float()
            print("MODBUS electric meter read: ", data)
            self.building_meter = round(data,3)
            client.close()

        except:
            print("ERROR ON MODBUS METER READ")
            client.close()
            self.building_meter = -1.0

        await asyncio.sleep(60)

async def main():
    bacnet_app = await BACnetApp.create()

    tasks = [bacnet_app.keep_baco_alive(), bacnet_app.modbus_meter_reader()]

    bacnet_app.server_check_lock = asyncio.Lock()  # Create a lock for synchronization

    await asyncio.gather(*tasks)

asyncio.run(main())
```

**Reply by bbartling** [reply]
I also forgot to note @rprronad  if you get some odd errors when updating the BACnet api for the Modbus sensor value i think it can because in BACnet they use very odd data types something like a primitive data type called a `Real`...

So you may have to do something like this below if you attempt to update the BACnet side with a numeric value and it errors out:

```
from bacpypes.primitivedata import Real


    async def update_bacnet_api(self, value):
        electric_meter_obj = self.bacnet.this_application.get_object_name("power-level")

        electric_meter_obj.presentValue = Real(value)
```

**Reply by bbartling** [reply]
This function looks sort of odd too but I notice that BAC0 dies if its not in a while loop...and if the asnycio sleeps are long like anything up to a second or greater i notice lag on the BACnet server when testing it out on a BACnet scan tool.
```
    async def keep_baco_alive(self):
        counter = 0
        while True:
            counter += 1
            await asyncio.sleep(0.01)
            if counter == 100:
                counter = 0
                async with self.server_check_lock:
                    await self.update_bacnet_api(self.building_meter)

```

**Comment by rprronad** [comment]
I really appreciate your patient explanations and quick responses and Thanks a lot `@bbartling` for sharing your knowledge and wishes.

**Reply by bbartling** [reply]
let me know how it goes...it maybe can be done better. Post whatever you make.

The Modbus side is a bit of a mystery so I would be learning from you on what you get working

**Reply by rprronad** [reply]
Okay sure @bbartling, 

& also If you got any updates, just mail me.
